Fix setUpClass() / tearDownClass() for device-specific tests

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151129

Finishes up the work started in #121686 + adds test

Update: this was not as straightforward as I originally imagined. Context below.

**TL;DR:** `TestFoo{CPU, CUDA}` now actually derive from `TestFoo`! Also, `{CPU, CUDA}TestBase` setup / teardown logic is now always called (it is required to set the primary device), regardless of whether `super().setUpClass()` / `super().tearDownClass()` are called or not.

**Background:** The typical way to get device-specific tests is to write a generic `TestFoo` and call `instantiate_device_type_tests(TestFoo, locals())` to get `TestFooCPU`, `TestFooCUDA`, etc. After this, generic tests (e.g. `TestFoo.test_bar()`) become `TestFooCPU.test_bar_cpu()` / `TestFooCUDA.test_bar_cuda()`.

Behind the scenes, this was historically accomplished by creating a `TestFooCUDA` that derives from both a `CUDATestBase` and an *empty class* called `TestFoo_base`. This `TestFoo_base` has the same bases as `TestFoo`, but none of the test functions (e.g. `test_bar()`). The documented reason for this is to avoid things like a derived `TestFooCUDA.test_bar()` being discovered in addition to the real device-specific test `TestFooCUDA.test_bar_cuda()`.

(1) A reason this matters is because it should be possible to call e.g. `super().setUpClass()` from a custom setup / teardown classmethod. If the generated TestFooCUDA does not derive from TestFoo, but instead derives from the empty class described above, this syntax does not work; in fact there is no way to form a proper `super()` call that works across the device-specific test variants. Here's an example that breaks in the OpInfo tests:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_ops.py#L218-L221

(2) Further, there is some precedent within a custom `setUpClass()` impl for storing things on the `cls` object to be accessed at test time. This must be the device-specific test class (`TestFooCUDA`) and not `TestFoo` for this to work. As an example, the open device registration tests load a module during setup and use it in the test logic:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L63-L77

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L79-L80

To accomplish both (1) and (2) at the same time, I decided to revisit the idea of utilizing a proper inheritance hierarchy for `TestFoo` -> `{TestFooCPU, TestFooCUDA}`. That is: have TestFooCPU / TestFooCUDA **actually** derive from `TestFoo`. This achieves both (1) and (2). The only thing left is to make sure the generic tests (e.g. `TestFoo.test_bar()`) are not discoverable, as was the stated reason for diverging from this in the first place. It turns out we can simply `delattr()` these generic tests from `TestFoo` once `TestFooCPU` / `TestFooCUDA` have been setup with the device-specific variants, and all works well. The `instantiate_device_type_tests(...)` logic already deletes `TestFoo` from scope, so I don't see a problem with deleting generic tests from this base class as well (CI will prove me right or wrong ofc).

**Side note:** I was encountering a weird race condition where sometimes the custom `setUpClass()` / `tearDownClass()` defined & swapped in [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L940-L955) would be used, and sometimes it wouldn't. This non-deterministic behavior was called out previously by @ngimel here:
https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/test/inductor/test_torchinductor_dynamic_shapes.py#L128-L130

To address this, I moved this block of logic to before the first call to `instantiate_test()`, as that method queries for the primary device, and the primary device identification logic may manually invoke `setUpClass()` (see [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L381-L384)). Goal: define the `setUpClass()` / `tearDownClass()` we want for correctness before they're ever called. This seems to work and the behavior is deterministic now AFAICT.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

jbschlosser committed:
Fix setUpClass() / tearDownClass() for device-specific tests

[ghstack-poisoned]

janeyx99 commented:
lol thanks!

jbschlosser committed:
Update on "Fix setUpClass() / tearDownClass() for device-specific tests"


Finishes up the work started in #121686 + adds test

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix setUpClass() / tearDownClass() for device-specific tests"


Finishes up the work started in #121686 + adds test

Update: this was not as straightforward as I originally imagined. Context below.

**Background:** The typical way to get device-specific tests is to write a generic `TestFoo` and call `instantiate_device_type_tests(TestFoo, locals())` to get `TestFooCPU`, `TestFooCUDA`, etc. After this, generic tests (e.g. `TestFoo.test_bar()`) become `TestFooCPU.test_bar_cpu()` / `TestFooCUDA.test_bar_cuda()`.

Behind the scenes, this was historically accomplished by creating a `TestFooCUDA` that derives from both a `CUDATestBase` and an *empty class* called `TestFoo_base`. This `TestFoo_base` has the same bases as `TestFoo`, but none of the test functions (e.g. `test_bar()`). The documented reason for this is to avoid things like a derived `TestFooCUDA.test_bar()` being discovered in addition to the real device-specific test `TestFooCUDA.test_bar_cuda()`.

(1) A reason this matters is because it should be possible to call e.g. `super().setUpClass()` from a custom setup / teardown classmethod. If the generated TestFooCUDA does not derive from TestFoo, but instead derives from the empty class described above, this syntax does not work; in fact there is no way to form a proper `super()` call that works across the device-specific test variants. Here's an example that breaks in the OpInfo tests:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_ops.py#L218-L221

(2) Further, there is some precedent within a custom `setUpClass()` impl for storing things on the `cls` object to be accessed at test time. This must be the device-specific test class (`TestFooCUDA`) and not `TestFoo` for this to work. As an example, the open device registration tests load a module during setup and use it in the test logic:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L63-L77

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L79-L80

To accomplish both (1) and (2) at the same time, I decided to revisit the idea of utilizing a proper inheritance hierarchy for `TestFoo` -> `{TestFooCPU, TestFooCUDA}`. That is: have TestFooCPU / TestFooCUDA **actually** derive from `TestFoo`. This achieves both (1) and (2). The only thing left is to make sure the generic tests (e.g. `TestFoo.test_bar()`) are not discoverable, as was the stated reason for diverging from this in the first place. It turns out we can simply `delattr()` these generic tests from `TestFoo` once `TestFooCPU` / `TestFooCUDA` have been setup with the device-specific variants, and all works well. The logic already deleted `TestFoo` from scope, so I don't see a problem with deleting generic tests from this base class as well (CI will prove me right or wrong ofc).

[ghstack-poisoned]

janeyx99 commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -930,37 +940,25 @@ def instantiate_device_type_tests(
                 else:
                     device_type_test_class.instantiate_test(name, copy.deepcopy(test))
-            else:  # Ports non-test member
-                assert (
```
so we don't want this assert anymore? what's the implication of no longer having the assert?

jbschlosser replied:
The new logic only brings over things from `generic_test_class` -> `device_type_test_class` that the latter doesn't already have.

In practice, `device_type_test_class` has these things starting out:
```
mappingproxy({'__module__': 'torch.testing._internal.common_device_type', '__doc__': None, '_classSetupFailed': False, '_class_cleanups': []})
```

These are all fine for the new device-specific class except arguably the `__module__`, and that it is set below (see line 952). So this just skips bringing these over from the generic test class.

janeyx99 commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -916,7 +903,30 @@ def instantiate_device_type_tests(
+            base.setUpClass()
+            # We want to call the @classmethod defined in the generic base, but pass
+            # it the device-specific class object (cls), hence the __func__ call.
+            generic_test_class.setUpClass.__func__(cls)
```
so are we able to call super here now?

jbschlosser replied:
not quite, this method conceptually is split into three parts:
1. Setup for `{CPU, CUDA}TestBase`; this sets the primary device, for example. This should always be called!
2. Setup for superclasses of `TestFoo`. This is optionally called, as defined by the author of `TestFoo`; they should make the `super()` call if they want this behavior.
3. Setup for `TestFoo` itself.

if we just call `super()` here, then we'll always do all three. But we want (2) to be optional as decided by the test case author; they decide whether or not to make the `super()` call in `TestFoo.setUpClass()`.

jbschlosser committed:
Update on "Fix setUpClass() / tearDownClass() for device-specific tests"


Finishes up the work started in #121686 + adds test

Update: this was not as straightforward as I originally imagined. Context below.

**TL;DR:** `TestFoo{CPU, CUDA}` now actually derive from `TestFoo`! Also, `{CPU, CUDA}TestBase` setup / teardown logic is now always called (it is required to set the primary device), regardless of whether `super().setUpClass()` / `super().tearDownClass()` are called or not.

**Background:** The typical way to get device-specific tests is to write a generic `TestFoo` and call `instantiate_device_type_tests(TestFoo, locals())` to get `TestFooCPU`, `TestFooCUDA`, etc. After this, generic tests (e.g. `TestFoo.test_bar()`) become `TestFooCPU.test_bar_cpu()` / `TestFooCUDA.test_bar_cuda()`.

Behind the scenes, this was historically accomplished by creating a `TestFooCUDA` that derives from both a `CUDATestBase` and an *empty class* called `TestFoo_base`. This `TestFoo_base` has the same bases as `TestFoo`, but none of the test functions (e.g. `test_bar()`). The documented reason for this is to avoid things like a derived `TestFooCUDA.test_bar()` being discovered in addition to the real device-specific test `TestFooCUDA.test_bar_cuda()`.

(1) A reason this matters is because it should be possible to call e.g. `super().setUpClass()` from a custom setup / teardown classmethod. If the generated TestFooCUDA does not derive from TestFoo, but instead derives from the empty class described above, this syntax does not work; in fact there is no way to form a proper `super()` call that works across the device-specific test variants. Here's an example that breaks in the OpInfo tests:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_ops.py#L218-L221

(2) Further, there is some precedent within a custom `setUpClass()` impl for storing things on the `cls` object to be accessed at test time. This must be the device-specific test class (`TestFooCUDA`) and not `TestFoo` for this to work. As an example, the open device registration tests load a module during setup and use it in the test logic:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L63-L77

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L79-L80

To accomplish both (1) and (2) at the same time, I decided to revisit the idea of utilizing a proper inheritance hierarchy for `TestFoo` -> `{TestFooCPU, TestFooCUDA}`. That is: have TestFooCPU / TestFooCUDA **actually** derive from `TestFoo`. This achieves both (1) and (2). The only thing left is to make sure the generic tests (e.g. `TestFoo.test_bar()`) are not discoverable, as was the stated reason for diverging from this in the first place. It turns out we can simply `delattr()` these generic tests from `TestFoo` once `TestFooCPU` / `TestFooCUDA` have been setup with the device-specific variants, and all works well. The `instantiate_device_type_tests(...)` logic already deletes `TestFoo` from scope, so I don't see a problem with deleting generic tests from this base class as well (CI will prove me right or wrong ofc).

**Side note:** I was encountering a weird race condition where sometimes the custom `setUpClass()` / `tearDownClass()` defined & swapped in [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L940-L955) would be used, and sometimes it wouldn't. To address this, I moved this block of logic to before the first call to `instantiate_test()`, as that method queries for the primary device, and the primary device identification logic may manually invoke `setUpClass()` (see [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L381-L384)). Goal: define the `setUpClass()` / `tearDownClass()` we want for correctness before they're ever called.

[ghstack-poisoned]

janeyx99 commented:
Alright I am happy to let CI be the final judge

jbschlosser commented:
`TestLinalg` does bad things that break after my changes!

https://github.com/pytorch/pytorch/blob/508b8825134e6d590f1fb495b2d2596e3689309e/test/test_linalg.py#L127-L133

`super(self.__class__, self)` is an anti-pattern in python will result in infinite recursion for derived classes (as described here). Replacing this nonsense with a normal call to `super()` should fix this.

jbschlosser committed:
Update on "Fix setUpClass() / tearDownClass() for device-specific tests"


Finishes up the work started in #121686 + adds test

Update: this was not as straightforward as I originally imagined. Context below.

**TL;DR:** `TestFoo{CPU, CUDA}` now actually derive from `TestFoo`! Also, `{CPU, CUDA}TestBase` setup / teardown logic is now always called (it is required to set the primary device), regardless of whether `super().setUpClass()` / `super().tearDownClass()` are called or not.

**Background:** The typical way to get device-specific tests is to write a generic `TestFoo` and call `instantiate_device_type_tests(TestFoo, locals())` to get `TestFooCPU`, `TestFooCUDA`, etc. After this, generic tests (e.g. `TestFoo.test_bar()`) become `TestFooCPU.test_bar_cpu()` / `TestFooCUDA.test_bar_cuda()`.

Behind the scenes, this was historically accomplished by creating a `TestFooCUDA` that derives from both a `CUDATestBase` and an *empty class* called `TestFoo_base`. This `TestFoo_base` has the same bases as `TestFoo`, but none of the test functions (e.g. `test_bar()`). The documented reason for this is to avoid things like a derived `TestFooCUDA.test_bar()` being discovered in addition to the real device-specific test `TestFooCUDA.test_bar_cuda()`.

(1) A reason this matters is because it should be possible to call e.g. `super().setUpClass()` from a custom setup / teardown classmethod. If the generated TestFooCUDA does not derive from TestFoo, but instead derives from the empty class described above, this syntax does not work; in fact there is no way to form a proper `super()` call that works across the device-specific test variants. Here's an example that breaks in the OpInfo tests:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_ops.py#L218-L221

(2) Further, there is some precedent within a custom `setUpClass()` impl for storing things on the `cls` object to be accessed at test time. This must be the device-specific test class (`TestFooCUDA`) and not `TestFoo` for this to work. As an example, the open device registration tests load a module during setup and use it in the test logic:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L63-L77

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L79-L80

To accomplish both (1) and (2) at the same time, I decided to revisit the idea of utilizing a proper inheritance hierarchy for `TestFoo` -> `{TestFooCPU, TestFooCUDA}`. That is: have TestFooCPU / TestFooCUDA **actually** derive from `TestFoo`. This achieves both (1) and (2). The only thing left is to make sure the generic tests (e.g. `TestFoo.test_bar()`) are not discoverable, as was the stated reason for diverging from this in the first place. It turns out we can simply `delattr()` these generic tests from `TestFoo` once `TestFooCPU` / `TestFooCUDA` have been setup with the device-specific variants, and all works well. The `instantiate_device_type_tests(...)` logic already deletes `TestFoo` from scope, so I don't see a problem with deleting generic tests from this base class as well (CI will prove me right or wrong ofc).

**Side note:** I was encountering a weird race condition where sometimes the custom `setUpClass()` / `tearDownClass()` defined & swapped in [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L940-L955) would be used, and sometimes it wouldn't. This non-deterministic behavior was called out previously by ngimel here:
https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/test/inductor/test_torchinductor_dynamic_shapes.py#L128-L130

To address this, I moved this block of logic to before the first call to `instantiate_test()`, as that method queries for the primary device, and the primary device identification logic may manually invoke `setUpClass()` (see [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L381-L384)). Goal: define the `setUpClass()` / `tearDownClass()` we want for correctness before they're ever called. This seems to work and the behavior is deterministic now AFAICT.

[ghstack-poisoned]

malfet commented on test/test_linalg.py:
```diff
@@ -125,12 +125,12 @@ def _hip_allow_tf32(self):
 
     def setUp(self):
-        super(self.__class__, self).setUp()
+        super().setUp()
```
Wow, that's a remnants of py2->py3-migration 
To be a bit pedantic, do you want to land it as BE change in separate PR?

jbschlosser replied:
> To be a bit pedantic, do you want to land it as BE change in separate PR?

I would split it but it's required to get CI passing sadly :/

malfet commented on test/test_testing.py:
```diff
@@ -440,6 +440,39 @@ def test_supported_dtypes(self, device, op):
 
+    def test_setup_and_teardown_run_for_device_specific_tests(self, device):
+        stderr = TestCase.runWithPytorchAPIUsageStderr(f"""\
+#!/usr/bin/env python3
+
+import torch
+from torch.testing._internal.common_device_type import instantiate_device_type_tests
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+class TestFoo(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        # store something on the test class to query during teardown
+        cls.stored_thing = "called with " + cls.__name__
+
+    @classmethod
+    def tearDownClass(cls):
+        # throw here so we know teardown was run
+        raise RuntimeError(cls.stored_thing)
+
+    def test_bar(self, device):
+        # make sure the test can access the stored thing
+        print(self.stored_thing)
+
+instantiate_device_type_tests(TestFoo, globals(), only_for='{self.device_type}')
+
+if __name__ == '__main__':
+    run_tests()
```
Nit, would be nice to keep those as separate files, say in `test/fixtures/`)

malfet commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -892,20 +892,7 @@ def instantiate_device_type_tests(
-    empty_class = type(empty_name, generic_test_class.__bases__, {})
-
-    # Acquires members names
-    # See Note [Overriding methods in generic tests]
```
Are you sure you want to remove reference to this note?

jbschlosser committed:
Update on "Fix setUpClass() / tearDownClass() for device-specific tests"


Finishes up the work started in #121686 + adds test

Update: this was not as straightforward as I originally imagined. Context below.

**TL;DR:** `TestFoo{CPU, CUDA}` now actually derive from `TestFoo`! Also, `{CPU, CUDA}TestBase` setup / teardown logic is now always called (it is required to set the primary device), regardless of whether `super().setUpClass()` / `super().tearDownClass()` are called or not.

**Background:** The typical way to get device-specific tests is to write a generic `TestFoo` and call `instantiate_device_type_tests(TestFoo, locals())` to get `TestFooCPU`, `TestFooCUDA`, etc. After this, generic tests (e.g. `TestFoo.test_bar()`) become `TestFooCPU.test_bar_cpu()` / `TestFooCUDA.test_bar_cuda()`.

Behind the scenes, this was historically accomplished by creating a `TestFooCUDA` that derives from both a `CUDATestBase` and an *empty class* called `TestFoo_base`. This `TestFoo_base` has the same bases as `TestFoo`, but none of the test functions (e.g. `test_bar()`). The documented reason for this is to avoid things like a derived `TestFooCUDA.test_bar()` being discovered in addition to the real device-specific test `TestFooCUDA.test_bar_cuda()`.

(1) A reason this matters is because it should be possible to call e.g. `super().setUpClass()` from a custom setup / teardown classmethod. If the generated TestFooCUDA does not derive from TestFoo, but instead derives from the empty class described above, this syntax does not work; in fact there is no way to form a proper `super()` call that works across the device-specific test variants. Here's an example that breaks in the OpInfo tests:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_ops.py#L218-L221

(2) Further, there is some precedent within a custom `setUpClass()` impl for storing things on the `cls` object to be accessed at test time. This must be the device-specific test class (`TestFooCUDA`) and not `TestFoo` for this to work. As an example, the open device registration tests load a module during setup and use it in the test logic:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L63-L77

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L79-L80

To accomplish both (1) and (2) at the same time, I decided to revisit the idea of utilizing a proper inheritance hierarchy for `TestFoo` -> `{TestFooCPU, TestFooCUDA}`. That is: have TestFooCPU / TestFooCUDA **actually** derive from `TestFoo`. This achieves both (1) and (2). The only thing left is to make sure the generic tests (e.g. `TestFoo.test_bar()`) are not discoverable, as was the stated reason for diverging from this in the first place. It turns out we can simply `delattr()` these generic tests from `TestFoo` once `TestFooCPU` / `TestFooCUDA` have been setup with the device-specific variants, and all works well. The `instantiate_device_type_tests(...)` logic already deletes `TestFoo` from scope, so I don't see a problem with deleting generic tests from this base class as well (CI will prove me right or wrong ofc).

**Side note:** I was encountering a weird race condition where sometimes the custom `setUpClass()` / `tearDownClass()` defined & swapped in [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L940-L955) would be used, and sometimes it wouldn't. This non-deterministic behavior was called out previously by ngimel here:
https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/test/inductor/test_torchinductor_dynamic_shapes.py#L128-L130

To address this, I moved this block of logic to before the first call to `instantiate_test()`, as that method queries for the primary device, and the primary device identification logic may manually invoke `setUpClass()` (see [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L381-L384)). Goal: define the `setUpClass()` / `tearDownClass()` we want for correctness before they're ever called. This seems to work and the behavior is deterministic now AFAICT.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot revert -m "flex attention tests failing"

jbschlosser commented:
@pytorchbot revert -m "flex attention tests failing" -c landrace

jbschlosser commented:
Land race with #144953, which does more bad things.

jbschlosser commented:
@pytorch rebase

jbschlosser commented:
@pytorchbot rebase

jbschlosser committed:
Update

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix setUpClass() / tearDownClass() for device-specific tests"


Finishes up the work started in #121686 + adds test

Update: this was not as straightforward as I originally imagined. Context below.

**TL;DR:** `TestFoo{CPU, CUDA}` now actually derive from `TestFoo`! Also, `{CPU, CUDA}TestBase` setup / teardown logic is now always called (it is required to set the primary device), regardless of whether `super().setUpClass()` / `super().tearDownClass()` are called or not.

**Background:** The typical way to get device-specific tests is to write a generic `TestFoo` and call `instantiate_device_type_tests(TestFoo, locals())` to get `TestFooCPU`, `TestFooCUDA`, etc. After this, generic tests (e.g. `TestFoo.test_bar()`) become `TestFooCPU.test_bar_cpu()` / `TestFooCUDA.test_bar_cuda()`.

Behind the scenes, this was historically accomplished by creating a `TestFooCUDA` that derives from both a `CUDATestBase` and an *empty class* called `TestFoo_base`. This `TestFoo_base` has the same bases as `TestFoo`, but none of the test functions (e.g. `test_bar()`). The documented reason for this is to avoid things like a derived `TestFooCUDA.test_bar()` being discovered in addition to the real device-specific test `TestFooCUDA.test_bar_cuda()`.

(1) A reason this matters is because it should be possible to call e.g. `super().setUpClass()` from a custom setup / teardown classmethod. If the generated TestFooCUDA does not derive from TestFoo, but instead derives from the empty class described above, this syntax does not work; in fact there is no way to form a proper `super()` call that works across the device-specific test variants. Here's an example that breaks in the OpInfo tests:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_ops.py#L218-L221

(2) Further, there is some precedent within a custom `setUpClass()` impl for storing things on the `cls` object to be accessed at test time. This must be the device-specific test class (`TestFooCUDA`) and not `TestFoo` for this to work. As an example, the open device registration tests load a module during setup and use it in the test logic:

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L63-L77

https://github.com/pytorch/pytorch/blob/070f3897453ea71a5505e4fcfaccb44e1e0cfa78/test/test_cpp_extensions_open_device_registration.py#L79-L80

To accomplish both (1) and (2) at the same time, I decided to revisit the idea of utilizing a proper inheritance hierarchy for `TestFoo` -> `{TestFooCPU, TestFooCUDA}`. That is: have TestFooCPU / TestFooCUDA **actually** derive from `TestFoo`. This achieves both (1) and (2). The only thing left is to make sure the generic tests (e.g. `TestFoo.test_bar()`) are not discoverable, as was the stated reason for diverging from this in the first place. It turns out we can simply `delattr()` these generic tests from `TestFoo` once `TestFooCPU` / `TestFooCUDA` have been setup with the device-specific variants, and all works well. The `instantiate_device_type_tests(...)` logic already deletes `TestFoo` from scope, so I don't see a problem with deleting generic tests from this base class as well (CI will prove me right or wrong ofc).

**Side note:** I was encountering a weird race condition where sometimes the custom `setUpClass()` / `tearDownClass()` defined & swapped in [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L940-L955) would be used, and sometimes it wouldn't. This non-deterministic behavior was called out previously by ngimel here:
https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/test/inductor/test_torchinductor_dynamic_shapes.py#L128-L130

To address this, I moved this block of logic to before the first call to `instantiate_test()`, as that method queries for the primary device, and the primary device identification logic may manually invoke `setUpClass()` (see [here](https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/testing/_internal/common_device_type.py#L381-L384)). Goal: define the `setUpClass()` / `tearDownClass()` we want for correctness before they're ever called. This seems to work and the behavior is deterministic now AFAICT.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
